src/datasets/utils/download_manager.py

```diff
@@ -182,6 +183,12 @@ def download(self, url_or_urls):
             downloaded_path(s): `str`, The downloaded paths matching the given input
                 url_or_urls.
         """
+        if self._download_config.only_splits:
+            if isinstance(url_or_urls, dict) and all(
+                split in url_or_urls for split in self._download_config.only_splits
+            ):
+                url_or_urls = {split: url_or_urls[split] for split in self._download_config.only_splits}
```

Comment:
That's a great trick ! However as you know this won't work on all cases unfortunately.
For example some datasets use a dictionary with keys that are not related to the splits since we allow any nested structure to be passed for downloading.

What do you think about adding an argument to `_split_generators` to tell it to return only a subset of the split generators ? In this case it would download only the files related to the requested splits

Then we can just edit the datasets that can benefit from this feature the most.

Reply:
The problem is that `split_generators` is already implemented in ALL canonical datasets without that parameter...

Reply:
Maybe we can pass the parameter only if it's part of the signature of `_split_generators` ?
We're already doing that for other optional parameters like `pipeline` in `BeamBasedBuilder._split_generators`:

https://github.com/huggingface/datasets/blob/909c58f6cf9568d73e87c89516a5e423983ad966/src/datasets/builder.py#L1065-L1067

I'm fine with this option but let me know if you have other ideas

Reply:
Now I pass the parameter to `_split_generators`. However, it must be handled properly by the method implementations.

Note that with my previous trick, it already works for `natural_questions` without even touching its current implementation of `split_generators.` 😉